editoast, front: forbid zero-length paths in pathfinding endpoint

Sh099078 opened the pull request:
Fixes https://github.com/OpenRailAssociation/osrd/issues/9444
- ~~Editoast: immediately return a `422 Unprocessable Content` error at deserialization when the endpoint `pathfinding/blocks` is called with the same origin and destination.~~
- Editoast: `pathfinding/blocks`: fail in post-processing with a new `PathfindingInputError::ZeroLengthPath` variant when core returns zero-length paths.
- Front: add a new type of stdcm error type when trying to make an STDCM request with the same origin and destination, update the stdcm inputs validation method.

emersion commented on editoast/src/views/path/pathfinding.rs:
```diff
@@ -73,6 +73,45 @@ struct PathfindingInput {
+            rolling_stock_maximum_speed,
+            rolling_stock_length,
+        } = Internal::deserialize(deserializer)?;
+        if path_items.first() == path_items.last() {
```
This is not enough to figure out if the path length will be zero. For instance, a UIC+ch could refer to the same OP as a trigram+ch.

Sh099078 replied:
I am missing some knowledge about what a trigram and an UIC represent and how to correlate them. Would you mind to elaborate or hint at some documentation on them ?

emersion replied:
Sure. There are multiple ways to refer to an operational point in a path: https://github.com/OpenRailAssociation/osrd/blob/9dedb376fde0341f4a6d6f94090dc1044c90d523/editoast/editoast_schemas/src/train_schedule/path_item.rs#L64

One way is via a _UIC code_. The UIC code is a worldwide-unique number which identifies a location. For instance, "87686006" is "Paris-Gare-de-Lyon". In France, the UIC is not enough to precisely identify an operational point: there are multiple operational points in "Paris-Gare-de-Lyon". So a _secondary code_ is added to pin down a specific one (often also called "ch" for "code chantier"). For instance, "BV" often refers to the place where the platforms are in a station ("bâtiment voyageur"). Side note, the UIC is made up of two parts, a country prefix ("87" for France) and the rest is called _CI_ ("code immuable", local French code, e.g. "686006").

Another way is via a _trigram_. It's a 3-letter code. It's essentially another way to write a CI, perhaps easier to remember for humans. For instance, "PLY" is the trigram for "Paris-Gare-de-Lyon".

Yet another way is via an _operational point ID_. It's just an opaque string identifier (e.g. "1e083f8b-c7db-48f4-bb12-ac3bb21d8963").

So, both of these refer to the same operational point:

```rust
OperationalPointDescription {
    trigram: "PLY",
    secondary_code: "BV",
}
OperationalPointUic {
    uic: 686006,
    secondary_code: "BV",
}
```

In other words, if these are given as input to the pathfinding algorithm, it'll freak out the same way it would if given two exact same path item locations.

Sh099078 replied:
Thank you for the explanation ! I added a new error variant `ZeroLengthPath` that transforms `core` success responses with zero-length paths into errors instead of checking whether the origin and destination are the same location before executing the pathfinding.

emersion commented on front/src/applications/stdcm/utils/checkStdcmConfigErrors.ts:
```diff
@@ -25,6 +25,10 @@ const checkStdcmConfigErrors = (
     return { errorType: StdcmConfigErrorTypes.PATHFINDING_FAILED };
   }
 
+  if (origin.name === destination.name && origin.ch === destination.ch) {
```
Would be more reliable to check the `uic` here instead of `name`: `name` is a human-readable string which may not be unique.

Sh099078 replied:
Thanks for the speedy review !

Fixed in https://github.com/OpenRailAssociation/osrd/commit/9cbbbaa531c152b60aca995288f3e8d5be94fe7d

emersion replied:
Looks good! Unfortunately we still have to write these `'uic' in …` checks, hopefully this can go away once Clara's refactorings are merged.

Sh099078 committed:
editoast, front: forbid zero-length paths

- Editoast: treat core pathfinding responses with a length of zero as
  errors.
- Frontend, stdcm: add error notifications when the user enters the same
  origin and destination.

Signed-off-by: Loup Federico <16464925+Sh099078@users.noreply.github.com>

woshilapin commented on editoast/src/views/path/pathfinding.rs:
```diff
@@ -448,6 +453,53 @@ pub mod tests {
+            .post(format!("/infra/{}/pathfinding/blocks", small_infra.id).as_str())
+            .json(&json!({
+                "path_items":[
+                {"trigram":"WS","secondary_code":"BV"},
+                {"trigram":"SWS","secondary_code":"BV"}
```
Hmm, as a reviewer of the code, I was expecting 2 identical path items, but I see 2 different trigrams and I'm left wondering why do we have a zero-length path. Isn't it possible to make the test even more obvious with the following?

```suggestion
                {"trigram":"WS","secondary_code":"BV"},
                {"trigram":"WS","secondary_code":"BV"}
```

emersion replied:
I think part of the point here is to ensure that the endpoint correctly returns the error even if the trigrams don't match?

Sh099078 replied:
Fixed in https://github.com/OpenRailAssociation/osrd/pull/9931/commits/ed12a878549fa1b8e8199791aeb95defd7a9ba87

woshilapin commented on front/public/locales/en/stdcm.json:
```diff
@@ -91,7 +91,8 @@
     "pathfindingFailed": "No path have been found for these waypoints.",
-    "requestFailed": "Calculation error for last minute train path"
+    "requestFailed": "Calculation error for last minute train path",
+    "sameOriginAndDestination": "The origin and destination cannot be the same."
```
nitpick: avoiding negation makes sometimes a clearer sentence?

```suggestion
    "sameOriginAndDestination": "The origin and destination must be different."
```

Sh099078 replied:
Fixed in https://github.com/OpenRailAssociation/osrd/pull/9931/commits/9719af2565978331c107cd86b909b6b3c5aab73b

woshilapin reviewed (approved):
Nice fix. Thank you.

emersion commented on editoast/src/core/pathfinding.rs:
```diff
@@ -133,6 +133,7 @@ pub enum PathfindingInputError {
     RollingStockNotFound {
         rolling_stock_name: String,
     },
+    ZeroLengthPath,
```
Each pathfinding error needs a translation in `front/public/locales/$lang/errors.json`.

emersion reviewed (commented):
Apart from this, looks good!

SharglutDev commented:
> [!NOTE]
> This will conflict with https://github.com/OpenRailAssociation/osrd/pull/9924

flomonster commented:
Should we authorize pathfinding that has a total length higher than `0m` but between two steps has a length of `0m`?

SharglutDev commented on front/src/applications/stdcm/utils/checkStdcmConfigErrors.ts:
```diff
@@ -30,6 +30,17 @@ const checkStdcmConfigErrors = (
   const isOriginRespectDestinationSchedule =
     origin.arrivalType === ArrivalTimeTypes.RESPECT_DESTINATION_SCHEDULE;
 
+  if (
+    'uic' in origin &&
+    'uic' in destination &&
+    'ch' in origin &&
+    'ch' in destination &&
+    origin.uic === destination.uic &&
+    origin.ch === destination.ch
+  ) {
```
We don't use `ch` in the code (too much sncf) but `secondary_code` (editoast type).
Both `uic` and `secondary_code` are defined if `location` is defined in `StdcmPathStep` type.
As we checked l.13 that we don't have any path step without location, we can assert that we will have one at this point.
```suggestion
 if (
    origin.location!.uic === destination.location!.uic &&
    origin.location!.secondary_code === destination.location!.secondary_code
  ) {
```

Sh099078 replied:
Fixed in https://github.com/OpenRailAssociation/osrd/commit/df926d1310b72c1d79879d4843d06ee2bf606ecb.

I also put the origin / destination check before the pathfinding response check as a bit of a hotfix, otherwise it would return the standard error when pathfinding fails instead of the new error that is more specific. It might be useful to give to `checkStdcmConfigErrors` more than just a boolean in case of pathfinding failure ? If it also received the error type returned by Editoast, we could directly return the  `zero_length_path` error instead of doing the current check on the front-end side.

SharglutDev replied:
Yep, you're totally right, it should be done [here](https://github.com/OpenRailAssociation/osrd/issues/9707) :)
But some errors/warning we want to display doesn't exist on editoast so it won't go away totally.

SharglutDev commented on front/src/applications/stdcm/utils/checkStdcmConfigErrors.ts:
We could go a little further and check in `useStaticPathfinding.ts` (probably going to conflict with 2 other PR atm) if origin and destination are the same to avoid launching the pathfinding. There is a check l.57 where we could do the same thing as here (maybe we can extract this check in a util like in stdcm/utils.index.ts)

Sh099078 replied:
Test added for same origin and destination in https://github.com/OpenRailAssociation/osrd/pull/9931/commits/30a4ecb19ddf9be2d69d59a9988eadb093a7632e

SharglutDev commented on front/src/applications/stdcm/hooks/useStaticPathfinding.ts:
Now that the pathfinding isn't launched when "zeroLengthPath", we don't get the warning below the form anymore. Indeed, in `StdcmConfig.tsx`, in the `useEffect` l.59, we update the warning only if we have a pathfinding but we don't anymore, so we should remove the if here.

Then we need to change the `checkStdcmConfigErrors` params a bit : if we leave it as it is, when the pathfinding actually works, you will see for a very short time the "pathfindingFailed" warning. This is because the `useEffect` is first updated when `pathSteps` are changing (we get the "pathfindingFailed" warning) and then the pathfinding starts and the warning disappears for the pathfinding loader.

To avoid that, in the function params, instead of passing `pathfindingStateError: boolean`, we can : 
- pass `pathfindingStatus?: 'success' | 'failure'` (at the end of the params because it can be undefined)
- change the `if` l.33 for `if (pathfindingStatus && pathfindingStatus === 'failure')`
- update the 2 spots where the function is called

Probably it would be better to make these changes in a separate commit.

Sh099078 replied:
Done in https://github.com/OpenRailAssociation/osrd/pull/9931/commits/a7b27cece696cc837a2b1e3a6188ad7142aaaf14

SharglutDev reviewed (commented):
Thank you for the changes, one more thing :)

Sh099078 committed:
front: display stdcm error message even when pathfinding is not called

Pathfinding is prevented when some validity checks don't pass, as for
example when the user selects the same origin and destination. The error
warning was only updated when pathfinging occurs, causing the user
warnings not to be displayed when we could detect before launching the
pathfinding that its request would be invalid.

Update the warnings independently on whether the pathfinding actually
happened so that invalid checks preventing the pathfinding also generate
their user warnings.

Signed-off-by: Loup Federico <16464925+Sh099078@users.noreply.github.com>

SharglutDev reviewed (approved):
Lgtm and tested, great job !
